Document assignment destructuring

simonbyrne opened the pull request:
This doesn't appear to be documented elsewhere in the manual.

simonbyrne committed:
Document assignment destructuring

JeffBezanson commented:
The feature is introduced a few lines below this, where it says `Julia supports simple tuple "destructuring" that facilitates this:`. We could just add a note there that the syntax actually works with all iterators. It is also called "destructuring assignment", not "assignment destructuring".

JeffBezanson commented on doc/src/manual/functions.md:
```diff
@@ -278,12 +278,34 @@ julia> x.a
+
+The iterator needs to be at least as long as the number of variables, and any excess elements
+are ignored. Additionally a single underscore `_` (which is an otherwise invalid variable name)
+can be used on the left to avoid assigning specific elements:
```
I think `_` should be explained elsewhere; it now seems like too many topics in this section. Maybe in variables or variables-and-scoping.

simonbyrne committed:
update based on comments

simonbyrne commented:
I've updated to address @JeffBezanson's coments (prompted by [discussion on discourse](https://discourse.julialang.org/t/destructuring-assignment-with-arrays-like-in-numpy/47441?u=simonbyrne)). 

I've moved some of the underscore discussion to variables, but it does occur to me that the whole destructuring issue is orthogonal to functions: ideally we would have an assignment section elsewhere (perhaps under variables or scoping?), but until then this seems to be the best place for it.

simonbyrne commented:
bump?

vtjnash commented:
I believe this is good, but now requires an update for #37410 also, which seems to have landed without docs. Would you be willing to take another pass through?

vtjnash commented on doc/src/manual/functions.md:
````diff
@@ -331,12 +331,36 @@ Named tuples are very similar to tuples, except that fields can additionally be
+on the left to avoid assigning specific elements:
+
+```jldoctest
+julia> _,_,_,d = 1:10
````
style nit: there should typically be spaces around operators and commas
```suggestion
julia> _, _, _, d = 1:10
```

vtjnash commented:
See also https://github.com/JuliaLang/julia/pull/31765

simonbyrne committed:
Update doc/src/manual/functions.md

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

simeonschaub commented on doc/src/manual/functions.md:
````diff
@@ -331,12 +331,36 @@ Named tuples are very similar to tuples, except that fields can additionally be
+4
+```
+
+This can be used to simulate returning multiple values from functions by returning a tuple or
+other iterable value. For example, the following function returns a two values:
````
```suggestion
This can be used to return multiple values from functions by returning a tuple or
other iterable value. For example, the following function returns two values:
```

simonbyrne committed:
underscores now mentioned elsewhere

simonbyrne committed:
Update doc/src/manual/functions.md

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

simonbyrne committed:
Clean up and add reference to slurping assigment (#37410)

simonbyrne committed:
whitespace

simonbyrne commented:
@vtjnash @simeonschaub Thanks for the comments, I have updated the PR.

simonbyrne committed:
remove spurious space

vtjnash commented:
This should perhaps also be combined with https://github.com/JuliaLang/julia/pull/30579

simonbyrne commented:
@vtjnash this is #30579

vtjnash commented:
How about https://github.com/JuliaLang/julia/pull/38125 then 😬?

simonbyrne committed:
mention swapping variables and other kinds of LHS

simonbyrne commented:
I added the points mentioned by @JeffBezanson in https://github.com/JuliaLang/julia/pull/38125#issuecomment-721910590.

simeonschaub commented on doc/src/manual/functions.md:
````diff
@@ -388,15 +401,71 @@ julia> y
+3
+```
+
+!!! compat "Julia 1.6"
+    `...` with assignment requires Julia 1.6
````
I think we usually put the compat annotations after the examples.

simonbyrne replied:
It seems a bit mixed: e.g. in `stacktraces.md` it is used at the start.

simeonschaub replied:
Ah, I see. Yeah, seems fine to me then.

simeonschaub commented on doc/src/manual/functions.md:
```diff
@@ -388,15 +401,71 @@ julia> y
+    `...` with assignment requires Julia 1.6
+
+If the last symbol in the assignment list is suffixed by `...` (known as _slurping_), then
+it will collect the remaining elements of the iterator:
```
```suggestion
it will collect the remaining elements of the iterator. For arbitrary iterators on the right
hand side, this will return a lazy iterator, but arrays, tuples and strings are collected
eagerly.
```

simonbyrne replied:
I edited this slightly, referring the reader to the `Base.rest` docs for precise behaviour.

simeonschaub replied:
Yeah, looks great!

simeonschaub commented on doc/src/manual/functions.md:
````diff
@@ -388,15 +401,71 @@ julia> y
+ 2
+ 3
+ 4
+```
````
````suggestion

julia> a, b... = Iterators.map(abs2, 1:4)
Base.Generator{UnitRange{Int64}, typeof(abs2)}(abs2, 1:4)

julia> a
1

julia> b
Base.Iterators.Rest{Base.Generator{UnitRange{Int64}, typeof(abs2)}, Int64}(Base.Generator{UnitRange{Int64}, typeof(abs2)}(abs2, 1:4), 1)

julia> collect(b)
3-element Vector{Int64}:
  4
  9
 16
```
````

simonbyrne replied:
This section is already quite long: perhaps we should move this to the doc string for `Base.rest`?

simeonschaub replied:
I wanted to emphasize here that slurping doesn't necessarily produce a vector and I don't expect many people to read the `Base.rest` doc string. Your changes look good to me.

simonbyrne replied:
Good point, I decided to keep most of it.

simeonschaub reviewed (approved):
Some small suggestions, but otherwise LGTM.

simonbyrne committed:
Address @simeonschaub's comments
